waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,605 @@
+                            {:name "MESOS_DIRECTORY" :value home-path}
+                            {:name "MESOS_SANDBOX" :value home-path}]
+                           (concat
+                             (for [[k v] common-env]
+                               {:name k :value v})
+                             (for [i (range ports)]
+                               {:name (str "PORT" i) :value (str (+ port0 i))})))
+        params {:k8s-name (service-id->k8s-name scheduler service-id)
+                :backend-protocol backend-proto
+                :backend-protocol-caps (string/upper-case backend-proto)
+                :cmd cmd
+                :cpus cpus
+                :env template-env
+                :grace-period-secs grace-period-secs
+                :health-check-interval-secs health-check-interval-secs
+                :health-check-max-consecutive-failures health-check-max-consecutive-failures
+                :health-check-url (sd/service-description->health-check-url service-description)
+                :home-path home-path
+                :memory  (str mem "Mi")
+                :min-instances min-instances
+                :port-count ports
+                :run-as-user run-as-user
+                :service-id service-id
+                :ssl? (= "https" backend-proto)}
+        edn-opts {:readers {'waiter/param params
+                            'waiter/param-str (comp str params)
+                            'waiter/port #(+ port0 %)
+                            'waiter.fn/into (fn [[xs ys]] (into xs ys))
+                            'waiter.fn/lower-case string/upper-case
+                            'waiter.fn/str #(apply str %)
+                            'waiter.fn/upper-case string/upper-case}}]
+    (try
+      (->> rs-spec-file-path slurp (edn/read-string edn-opts))
+      (catch Throwable e
+        (log/error e "Error creating ReplicaSet specification for" service-id)
+        (throw e)))))
+
+(defn- create-service
+  "Reify a Waiter Service as a Kuberentes ReplicaSet."
+  [service-id descriptor {:keys [api-server-url http-client] :as scheduler} service-id->password-fn]
+  (let [{:strs [run-as-user] :as service-description} (:service-description descriptor)
+        spec-json (service-spec scheduler service-id service-description service-id->password-fn)
+        request-url (str api-server-url
+                         "/apis/extensions/v1beta1/namespaces/"
+                         (service-description->namespace service-description)
+                         "/replicasets")
+        response-json (api-request http-client request-url
+                                   :body (as-json spec-json)
+                                   :request-method :post)]
+    (when-not (= "ReplicaSet" (:kind response-json))
+      (log/error "Invalid response from ReplicaSet create request"
+                 (as-json response-json)))
```

Comment:
Why convert the response back to a json string in the log?

Reply:
I can't remember why I did that... Deleted the `as-json` call.

Reply:
Actually, I don't know why I added this check at all. Kubernetes would never return a different kind of object when we're posting to the repicasets endpoint—and if it did, we'd want to abort, not log the error and continue... I'm deleting this whole check.